File: 02_Data_Structures/linked_lists/circular_linked_list/implementations/operations.py
# operations.py
# Circular Linked List – Core Operations in Python
#
# Supported operations:
# 1. insert_head
# 2. insert_end
# 3. delete_val
# 4. search
# 5. display
#
# A circular linked list connects the last node back to the head.
# We maintain only 'head' pointer to keep implementation simple.
#
# Time Complexity:
# insert_head: O(n) (to find last node)
# insert_end: O(n)
# delete_val: O(n)
# search: O(n)
# display: O(n)


class Node:
    def __init__(self, val):
        self.data = val
        self.next = None


class CircularLinkedList:
    def __init__(self):
        self.head = None

    def insert_head(self, val):
        new = Node(val)

        if self.head is None:
            self.head = new
            self.head.next = self.head
            return

        temp = self.head
        while temp.next != self.head:
            temp = temp.next

        new.next = self.head
        temp.next = new
        self.head = new

    def insert_end(self, val):
        new = Node(val)

        if self.head is None:
            self.head = new
            self.head.next = self.head
            return

        temp = self.head
        while temp.next != self.head:
            temp = temp.next

        temp.next = new
        new.next = self.head

    def delete_val(self, val):
        if self.head is None:
            return

        # Single node case
        if self.head.data == val and self.head.next == self.head:
            self.head = None
            return

        curr = self.head
        prev = None

        while curr.data != val:
            prev = curr
            curr = curr.next
            if curr == self.head:
                return  # not found

        # delete head
        if curr == self.head:
            tail = self.head
            while tail.next != self.head:
                tail = tail.next
            self.head = self.head.next
            tail.next = self.head
            return

        prev.next = curr.next

    def search(self, val):
        if self.head is None:
            return False

        temp = self.head
        while True:
            if temp.data == val:
                return True
            temp = temp.next
            if temp == self.head:
                break

        return False

    def display(self):
        if self.head is None:
            print("List is empty")
            return

        temp = self.head
        while True:
            print(temp.data, end=" ")
            temp = temp.next
            if temp == self.head:
                break
        print()


# Example usage
if __name__ == "__main__":
    cl = CircularLinkedList()
    cl.insert_head(3)
    cl.insert_head(2)
    cl.insert_end(4)
    cl.insert_end(5)

    cl.display()  # 2 3 4 5

    cl.delete_val(3)
    cl.display()  # 2 4 5

    print("Search 4:", cl.search(4))
    print("Search 10:", cl.search(10))

File: 02_Data_Structures/linked_lists/circular_linked_list/implementations/operations.cpp
// operations.cpp
// Circular Linked List – Core Operations Implementation
//
// This file implements basic operations on a Circular Singly Linked List:
// 1. Insert at head
// 2. Insert at end
// 3. Delete a node
// 4. Search
// 5. Display
//
// Explanation:
// A circular linked list is a linked list where the last node points back to the head.
// We maintain at minimum a pointer to 'head'. For convenience, we may also maintain 'tail'.
//
// Complexity:
// Insert at head: O(1)
// Insert at end: O(1) if tail maintained, else O(n)
// Delete: O(n)
// Search: O(n)
// Display: O(n)

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node(int v) : data(v), next(nullptr) {}
};

class CircularLinkedList {
private:
    Node* head;

public:
    CircularLinkedList() : head(nullptr) {}

    // Insert node at beginning
    void insertHead(int val) {
        Node* n = new Node(val);

        if (!head) {
            head = n;
            head->next = head;
            return;
        }

        Node* temp = head;
        while (temp->next != head) temp = temp->next;

        // temp is last node
        n->next = head;
        temp->next = n;
        head = n;
    }

    // Insert at end
    void insertEnd(int val) {
        Node* n = new Node(val);

        if (!head) {
            head = n;
            head->next = head;
            return;
        }

        Node* temp = head;
        while (temp->next != head) temp = temp->next;

        temp->next = n;
        n->next = head;
    }

    // Delete a node with value
    void deleteVal(int val) {
        if (!head) return;

        // If only 1 node
        if (head->data == val && head->next == head) {
            delete head;
            head = nullptr;
            return;
        }

        Node* curr = head;
        Node* prev = nullptr;

        // Find node
        while (curr->data != val) {
            prev = curr;
            curr = curr->next;

            if (curr == head) return; // not found
        }

        // Delete head case
        if (curr == head) {
            Node* temp = head;
            while (temp->next != head) temp = temp->next;
            head = head->next;
            temp->next = head;
            delete curr;
            return;
        }

        // Delete non-head node
        prev->next = curr->next;
        delete curr;
    }

    // Search a value
    bool search(int val) {
        if (!head) return false;

        Node* temp = head;

        do {
            if (temp->data == val) return true;
            temp = temp->next;
        } while (temp != head);

        return false;
    }

    // Display list
    void display() {
        if (!head) {
            cout << "List is empty\n";
            return;
        }

        Node* temp = head;

        do {
            cout << temp->data << " ";
            temp = temp->next;
        } while (temp != head);

        cout << "\n";
    }
};

int main() {
    CircularLinkedList cl;
    cl.insertHead(3);
    cl.insertHead(2);
    cl.insertEnd(4);
    cl.insertEnd(5);

    cl.display(); // 2 3 4 5

    cl.deleteVal(3);
    cl.display(); // 2 4 5

    cout << (cl.search(4) ? "Found" : "Not found") << "\n";

    return 0;
}

